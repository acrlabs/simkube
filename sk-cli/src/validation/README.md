# SimKube Trace Validation Checks

| code | name | description |
|---|---|---|
| W0000 | status_field_populated |  Indicates that the status field of a Kubernetes object in the trace is non-empty; status fields are updated by their controlling objects and shouldn't be applied "by hand".  This is probably "fine" but it would be better to clean them up (and also they take up a lot of space.  |

This file is auto-generated; to rebuild, run `make validation_rules`.
