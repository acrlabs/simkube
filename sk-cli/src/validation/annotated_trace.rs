use std::collections::BTreeMap;

use sk_core::external_storage::{
    ObjectStoreWrapper,
    SkObjectStore,
};
use sk_store::{
    TraceEvent,
    TraceStorable,
    TraceStore,
};

#[derive(Clone, Default)]
pub struct AnnotatedTraceEvent {
    pub data: TraceEvent,
    pub annotations: Vec<(usize, String)>,
}

#[derive(Default)]
pub struct AnnotatedTrace {
    pub base: TraceStore,
    pub path: String,
    pub events: Vec<AnnotatedTraceEvent>,
    pub summary: BTreeMap<String, usize>,
}

impl AnnotatedTrace {
    pub async fn new(trace_path: &str) -> anyhow::Result<AnnotatedTrace> {
        let object_store = SkObjectStore::new(trace_path)?;
        let trace_data = object_store.get().await?.to_vec();
        let base = TraceStore::import(trace_data, &None)?;
        let events = base
            .iter()
            .map(|(event, _)| AnnotatedTraceEvent { data: event.clone(), ..Default::default() })
            .collect();
        Ok(AnnotatedTrace {
            base,
            events,
            path: trace_path.into(),
            ..Default::default()
        })
    }
}
