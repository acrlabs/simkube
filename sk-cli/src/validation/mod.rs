mod annotated_trace;
mod status_field_populated;
mod validation_store;
mod validator;

use clap::{
    value_parser,
    Subcommand,
    ValueEnum,
};
use sk_core::prelude::*;

pub use self::annotated_trace::AnnotatedTrace;
pub use self::validation_store::ValidationStore;

#[derive(Subcommand)]
pub enum ValidateSubcommand {
    #[command(about = "check a trace file")]
    Check(CheckArgs),

    #[command(about = "explain a rule")]
    Explain(ExplainArgs),

    #[command(about = "print all validation rules")]
    Print(PrintArgs),
}

#[derive(clap::Args)]
pub struct CheckArgs {
    #[arg(long_help = "location of the input trace file")]
    pub trace_path: String,
}

#[derive(clap::Args)]
pub struct ExplainArgs {
    #[arg(long_help = "Error code to explain")]
    pub code: String,
}

#[derive(Clone, ValueEnum)]
pub enum PrintFormat {
    Json,
    List,
    Table,
    Yaml,
}

#[derive(clap::Args)]
pub struct PrintArgs {
    #[arg(
        short,
        long,
        long_help = "format to display the validation rules",
        default_value = "list",
        value_parser = value_parser!(PrintFormat),
    )]
    pub format: PrintFormat,
}

pub async fn cmd(subcommand: &ValidateSubcommand) -> EmptyResult {
    let mut validators = ValidationStore::default();
    match subcommand {
        ValidateSubcommand::Check(args) => {
            let mut trace = AnnotatedTrace::new(&args.trace_path).await?;
            validators.validate_trace(&mut trace);
            print_summary(&trace, &validators)?;
        },
        ValidateSubcommand::Explain(args) => validators.explain(&args.code)?,
        ValidateSubcommand::Print(args) => validators.print(&args.format)?,
    }
    Ok(())
}

fn print_summary(trace: &AnnotatedTrace, validators: &ValidationStore) -> EmptyResult {
    for (code, count) in trace.summary_iter() {
        let name = validators.lookup(code)?.name;
        println!("{name} ({code}): {count:.>30}");
    }
    Ok(())
}

#[cfg(test)]
pub mod tests;
