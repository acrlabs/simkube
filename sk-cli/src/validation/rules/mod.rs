pub mod service_account_missing;
pub mod status_field_populated;

#[cfg(test)]
mod tests;
