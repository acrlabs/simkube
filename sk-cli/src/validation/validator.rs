use std::fmt;

use serde::{
    Serialize,
    Serializer,
};

use super::annotated_trace::AnnotatedTraceEvent;

#[derive(Eq, Hash, PartialEq, Serialize)]
pub enum ValidatorType {
    Warning,
    #[allow(dead_code)]
    Error,
}

impl fmt::Display for ValidatorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                ValidatorType::Warning => 'W',
                ValidatorType::Error => 'E',
            }
        )
    }
}

pub trait Diagnostic {
    fn check_next_event(&mut self, evt: &mut AnnotatedTraceEvent) -> Vec<usize>;
    fn reset(&mut self);
}

#[derive(Serialize)]
pub struct Validator {
    #[serde(rename = "type")]
    pub type_: ValidatorType,
    pub name: &'static str,

    #[serde(serialize_with = "flatten_str")]
    pub help: &'static str,

    #[serde(skip)]
    pub diagnostic: Box<dyn Diagnostic>,
}

impl Validator {
    pub fn check_next_event(&mut self, a_event: &mut AnnotatedTraceEvent) -> Vec<usize> {
        self.diagnostic.check_next_event(a_event)
    }

    pub fn reset(&mut self) {
        self.diagnostic.reset()
    }

    pub fn help(&self) -> String {
        self.help.replace('\n', " ")
    }
}

fn flatten_str<S: Serializer>(s: &str, ser: S) -> Result<S::Ok, S::Error> {
    ser.serialize_str(&s.replace('\n', " "))
}
