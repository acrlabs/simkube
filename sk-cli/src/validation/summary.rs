use std::collections::BTreeMap; // BTreeMap sorts by key, HashMap doesn't
use std::fmt;

use super::validator::ValidatorCode;
use super::VALIDATORS;

const WIDTH: usize = 70;

#[derive(Default)]
pub struct ValidationSummary {
    pub annotations: BTreeMap<ValidatorCode, usize>,
    pub patches: usize,
}

impl fmt::Display for ValidationSummary {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "\nValidation errors found:")?;
        writeln!(f, "{}", "-".repeat(WIDTH))?;
        for (code, count) in self.annotations.iter() {
            if *count == 0 {
                continue;
            }
            let name = VALIDATORS.lookup(code).map(|v| v.name).unwrap_or("<unknown>");
            let left = format!("{name} ({code})");
            let right = format!("{count}");
            let mid_width = WIDTH.saturating_sub(left.len()).saturating_sub(right.len()).saturating_sub(2); // two chars for extra spaces
            writeln!(f, "{left} {} {right}", ".".repeat(mid_width))?;
        }

        if self.patches > 0 {
            writeln!(f, "{}", "-".repeat(WIDTH))?;
            writeln!(f, "Patches applied: {}", self.patches)?;
            writeln!(f, "0 problems remaining")?;
        }
        Ok(())
    }
}
