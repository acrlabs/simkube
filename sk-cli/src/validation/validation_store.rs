use std::collections::BTreeMap; // BTreeMap sorts by key, HashMap doesn't

use anyhow::anyhow;
use lazy_static::lazy_static;
use serde::Serialize;
use sk_core::prelude::*;

use super::summary::ValidationSummary;
use super::validator::{
    Validator,
    ValidatorCode,
};
use super::{
    status_field_populated,
    AnnotatedTrace,
    AnnotatedTracePatch,
    PatchLocations,
    PrintFormat,
};

#[derive(Serialize)]
pub struct ValidationStore {
    pub(super) validators: BTreeMap<ValidatorCode, Validator>,
}

impl ValidationStore {
    pub fn validate_trace(&self, trace: &mut AnnotatedTrace, fix: bool) -> anyhow::Result<ValidationSummary> {
        for validator in self.validators.values() {
            validator.reset();
        }

        let mut summary = ValidationSummary::default();
        let mut summary_populated = false;
        loop {
            let s = trace.validate(&self.validators);
            if !summary_populated {
                summary.annotations = s;
                summary_populated = true;
            }

            if !fix {
                break;
            }

            let Some(next_error) = trace.get_next_error() else {
                break;
            };

            let Some(op) = self
                .validators
                .get(&next_error)
                .ok_or(anyhow!("validation error"))?
                .fixes()
                .first()
                .cloned()
            else {
                println!("no fix available for {next_error}; continuing");
                break;
            };
            summary.patches += trace.apply_patch(AnnotatedTracePatch {
                locations: PatchLocations::AffectedObjects(next_error),
                op,
            })?;
        }

        Ok(summary)
    }

    pub(super) fn explain(&self, code: &ValidatorCode) -> EmptyResult {
        let v = self.lookup(code)?;
        println!("{} ({code})", v.name);
        println!("{:=<80}", "");
        println!("{}", v.help);
        Ok(())
    }

    pub(super) fn lookup<'a>(&'a self, code: &ValidatorCode) -> anyhow::Result<&'a Validator> {
        self.validators.get(code).ok_or(anyhow!("code not found: {code}"))
    }

    pub(super) fn print(&self, format: &PrintFormat) -> EmptyResult {
        match format {
            PrintFormat::Json => print!("{}", serde_json::to_string(self)?),
            PrintFormat::List => self.print_list()?,
            PrintFormat::Table => self.print_table()?,
            PrintFormat::Yaml => print!("{}", serde_yaml::to_string(self)?),
        }

        Ok(())
    }

    fn print_list(&self) -> EmptyResult {
        for (code, validator) in self.validators.iter() {
            println!("{} ({}): {}", code, validator.name, validator.help());
        }
        Ok(())
    }

    fn print_table(&self) -> EmptyResult {
        println!("| code | name | description |");
        println!("|---|---|---|");
        for (code, validator) in self.validators.iter() {
            println!("| {} | {} | {} |", code, validator.name, validator.help());
        }
        Ok(())
    }

    fn new() -> ValidationStore {
        let mut store = ValidationStore { validators: BTreeMap::new() };

        store.register(status_field_populated::validator());

        store
    }

    fn register(&mut self, v: Validator) {
        let code = ValidatorCode(v.type_, self.validators.len());
        self.register_with_code(code, v);
    }

    fn register_with_code(&mut self, code: ValidatorCode, v: Validator) {
        self.validators.insert(code, v);
    }
}

lazy_static! {
    pub static ref VALIDATORS: ValidationStore = ValidationStore::new();
}
