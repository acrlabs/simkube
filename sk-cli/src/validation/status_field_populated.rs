use super::annotated_trace::AnnotatedTraceEvent;
use super::validation_store::{
    Diagnostic,
    Validator,
    ValidatorType,
};

const HELP: &str = r#"Indicates that the status field of a Kubernetes object in
the trace is non-empty; status fields are updated by their controlling objects
and shouldn't be applied "by hand".  This is probably "fine" but it would be
better to clean them up (and also they take up a lot of space."#;

#[derive(Default)]
pub(super) struct StatusFieldPopulated {}

impl Diagnostic for StatusFieldPopulated {
    fn check_next_event(&mut self, event: &mut AnnotatedTraceEvent) -> Vec<usize> {
        event
            .data
            .applied_objs
            .iter()
            .enumerate()
            .filter_map(|(i, obj)| {
                if obj
                    .data
                    .as_object()
                    .expect("DynamicObject data should be a map")
                    .get("status")
                    .is_some_and(|v| !v.is_null())
                {
                    return Some(i);
                }
                None
            })
            .collect()
    }

    fn reset(&mut self) {}
}

pub(super) fn validator() -> Validator {
    Validator {
        type_: ValidatorType::Warning,
        name: "status_field_populated",
        help: HELP,
        diagnostic: Box::new(StatusFieldPopulated::default()),
    }
}
