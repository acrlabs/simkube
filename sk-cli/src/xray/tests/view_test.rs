use insta::assert_debug_snapshot;
use ratatui::backend::TestBackend;
use ratatui::prelude::*;
use ratatui::widgets::ListState;

use super::*;
use crate::validation::tests::annotated_trace;
use crate::validation::AnnotatedTrace;

#[fixture]
fn test_app(annotated_trace: AnnotatedTrace) -> App {
    App {
        annotated_trace,
        event_list_state: ListState::default().with_selected(Some(0)),
        ..Default::default()
    }
}

#[rstest]
#[case::first(0)]
#[case::last(3)]
fn itest_render_event_list(mut test_app: App, #[case] index: usize) {
    set_snapshot_suffix!("{index}");
    test_app.event_list_state.select(Some(index));
    let mut term = Terminal::new(TestBackend::new(80, 20)).unwrap();
    let cf = term.draw(|frame| view(&mut test_app, frame)).unwrap();
    assert_debug_snapshot!(cf);
}

#[rstest]
#[case::first(0)]
#[case::middle(2)]
#[case::last(3)]
fn itest_render_event_list_event_selected(mut test_app: App, #[case] index: usize) {
    set_snapshot_suffix!("{index}");
    test_app.mode = Mode::EventSelected;
    test_app.event_list_state.select(Some(index));
    test_app.object_list_state.select(Some(0));
    let mut term = Terminal::new(TestBackend::new(80, 20)).unwrap();
    let cf = term.draw(|frame| view(&mut test_app, frame)).unwrap();
    assert_debug_snapshot!(cf);
}
