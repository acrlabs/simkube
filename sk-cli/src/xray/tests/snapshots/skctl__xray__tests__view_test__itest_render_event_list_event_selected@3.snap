---
source: sk-cli/src/xray/tests/view_test.rs
expression: cf
---
CompletedFrame {
    buffer: Buffer {
        area: Rect { x: 0, y: 0, width: 80, height: 20 },
        content: [
            "┌──────────────────────────────────────────────────────────────────────────────┐",
            "│   00:00:00 (0 applied/0 deleted)                                             │",
            "│   00:00:01 (1 applied/0 deleted)                                             │",
            "│   00:00:02 (2 applied/0 deleted)                                             │",
            "│>> 00:00:03 (0 applied/1 deleted)                                             │",
            "│++   - test-namespace/test_depl1                                              │",
            "│                                                                              │",
            "│                                                                              │",
            "│                                                                              │",
            "│                                                                              │",
            "│                                                                              │",
            "│                                                                              │",
            "│                                                                              │",
            "│                                                                              │",
            "└──────────────────────────────────────────────────────────────────────────────┘",
            "┌──────────────────────────────────────────────────────────────────────────────┐",
            "│Hello SimKube!                                                                │",
            "│Use arrows to navigate, space to select, 'q' to quit.                         │",
            "│                                                                              │",
            "└──────────────────────────────────────────────────────────────────────────────┘",
        ],
        styles: [
            x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 1, y: 4, fg: Reset, bg: Reset, underline: Reset, modifier: REVERSED,
            x: 79, y: 4, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 1, y: 5, fg: Reset, bg: Blue, underline: Reset, modifier: ITALIC,
            x: 79, y: 5, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 0, y: 15, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        ]
    },
    area: Rect {
        x: 0,
        y: 0,
        width: 80,
        height: 20,
    },
    count: 0,
}
